pylabrobot/pumps/calibration_tests.py

```diff
@@ -0,0 +1,54 @@
+import unittest
+
+from pylabrobot.pumps.calibration import PumpCalibration
+
+import pylabrobot
+import os
+
+plr_directory = os.path.join(pylabrobot.__path__[0], "testing", "test_data")
+
+
+class TestCalibration(unittest.TestCase):
+  """
+    Tests for the PumpCalibration class.
+  """
+
+  def setUp(self) -> None:
+    self.json_list_path = os.path.join(plr_directory, "test_calibration_list.json")
+    self.json_dict_path = os.path.join(plr_directory, "test_calibration_dict.json")
+    self.json_null_path = os.path.join(plr_directory, "test_calibration_null.json")
+    self.csv_path = os.path.join(plr_directory, "test_calibration.csv")
+    self.txt_path = os.path.join(plr_directory, "test_calibration.txt")
+
+  def test_load_calibration(self):
+    calibration = PumpCalibration.load_calibration()
+    self.assertIsNone(calibration)
+    calibration = PumpCalibration.load_calibration([1.0, 2.0])
+    self.assertEqual(calibration[0], 1.0)
+    self.assertEqual(calibration[1], 2.0)
+    calibration = PumpCalibration.load_calibration({0: 1.0, 1: 2.0})
+    self.assertEqual(calibration[0], 1.0)
+    self.assertEqual(calibration[1], 2.0)
+    calibration = PumpCalibration.load_calibration({0: 1.0, 1: 2.0})
+    self.assertEqual(calibration[0], 2.0)
+    self.assertEqual(calibration[1], 1.0)
+    self.assertRaises(ValueError, PumpCalibration.load_calibration, {0: 1.0, 1: 2.0, 3: 3.0})
+    self.assertRaises(ValueError, PumpCalibration.load_calibration, {0: -1.0, 1: 2.0, 2: 3.0})
+    self.assertRaises(ValueError, PumpCalibration.load_calibration, {-1: 1.0, 0: 1.0, 1: 2.0})
+    self.assertRaises(ValueError, PumpCalibration.load_calibration, {0: 1.0, 1: 1.0, 2: 2.0})
+    calibration = PumpCalibration.load_calibration(1.0, num_items=2)
+    self.assertEqual(calibration[0], 1.0)
+    self.assertEqual(calibration[1], 1.0)
+    calibration = PumpCalibration.load_calibration(self.csv_path)
+    self.assertEqual(calibration[0], 1.0)
+    self.assertEqual(calibration[1], 2.0)
+    calibration = PumpCalibration.load_calibration(self.json_list_path)
+    self.assertEqual(calibration[0], 1)
+    self.assertEqual(calibration[1], 1)
+    calibration = PumpCalibration.load_calibration(self.json_dict_path)
+    self.assertEqual(calibration[0], 1.0)
+    self.assertEqual(calibration[1], 1.0)
+    self.assertRaises(TypeError, PumpCalibration.load_calibration, self.json_null_path)
+    self.assertRaises(NotImplementedError,
+                      PumpCalibration.load_calibration,
+                      self.txt_path)
```

Comment:
I think these could be split up in separate tests for loading from a list, dictionary, csv, etc. This makes it easier to see where the issue is when a tests fails (based on which tests pass/fail), and allow testing of individual methods (tests should run very quickly here so not really an issue, but it's good practice to split it up).

Reply:
I ran into this problem when I was debugging. I think I want to keep a full set of tests like this for load_calibration, but only for the ones where there is some transformation of the input prior to the loading into the sub-function. I'll also add the dictionary and lists to SetUp so they don't have to be written out here.